beancount/plugins/close_tree.py

```diff
@@ -0,0 +1,98 @@
+closed. Unopened parent accounts can also be closed. Any explicitly specified close is left
+untouched.
+
+For example, given this:
+
+```
+2017-11-10 open Assets:Brokerage:AAPL
+2017-11-10 open Assets:Brokerage:ORNG
+2018-11-10 close Assets:Brokerage  ; this account does not necessarily need to be opened
+```
+
+the plugin turns it into:
+```
+2017-11-10 open Assets:Brokerage:AAPL
+2017-11-10 open Assets:Brokerage:ORNG
+2018-11-10 close Assets:Brokerage:AAPL
+2018-11-10 close Assets:Brokerage:ORNG
+```
+
+Invoke this plugin _after_ any plugins that generate `open` directives for account trees that you
+want to auto close. An example is the `auto_accounts` plugin that ships with Beancount:
+
+```
+plugin "beancount.plugins.auto_accounts"
+plugin "beancount.plugins.close_tree"
+```
+"""
+
+import time
+from beancount.core import data
+from beancount.core.data import Open, Close
+
+DEBUG = 0
+__plugins__ = ('close_tree',)
+
+
+def close_tree(entries, unused_options_map):
+    """Insert close entries for all subaccounts of a closed account.
+
+    Args:
+      entries: A list of directives. We're interested only in the Open/Close instances.
+      unused_options_map: A parser options dict.
+    Returns:
+      A tuple of entries and errors. """
+
+    start_time = time.time()
+    close_count = 0
+    new_entries = []
+    errors = []
+
+    opens = [e.account for e in entries if isinstance(e, Open)]
+    closes = [e.account for e in entries if isinstance(e, Close)]
+
+    for entry in entries:
+        if isinstance(entry,  Close):
+            subaccounts = [a for a in opens if entry.account + ':' in a and a not in closes]
+            for s in subaccounts:
+                meta = data.new_metadata('<beancount.plugins.close_tree>', 0)
+                close_entry = data.Close(meta, entry.date, s)
+                new_entries.append(close_entry)
+                closes.append(s)
+                close_count += 1
+            if entry.account in opens:
+                new_entries.append(entry)
+        else:
+            new_entries.append(entry)
+
+    # beancount.plugins.auto_accounts, if used in conjunction with this plugin, inserts an Open
+    # entry for empty parent accounts which we are trying to close. For example:
+
+    # plugin "beancount.plugins.auto_accounts"
+    # 2021-06-17 close Expenses:Non-Retirement:Auto:Fit
+    # plugin "beancount.plugins.close_tree"
+    # 2019-01-01 * "Transaction"
+    #   Expenses:Non-Retirement:Auto:Fit:Insurance      10 USD
+    #   Expenses:Non-Retirement:Auto:Fit:Gas            20 USD
+    #   Assets:Transfer
+
+    # Above, auto_accounts inserts an Open for 'Expenses:Non-Retirement:Auto:Fit:Gas' on 2021-06-17
+    # since it appears in a close directive. But that's not what we want. The semantic of the close
+    # when u used with this plugin is: "close all children, and close the parent if it was opened
+    # earlier. Below, we filter out entries to correctly handle this scenario:
```

Comment:
I had a very hard time to follow what you are doing here. I had to unpack the different steps. Starting with this ledger:
```
2019-01-01 * "Transaction"
  Expenses:Non-Retirement:Auto:Fit:Insurance      10 USD
  Expenses:Non-Retirement:Auto:Fit:Gas            20 USD
  Assets:Transfer
2021-06-17 close Expenses:Non-Retirement:Auto:Fit
```
the `bencount.plugins.auto_accounts` adds the `open` directives:
```
2019-01-01 open Assets:Transfer
2019-01-01 open Expenses:Non-Retirement:Auto:Fit:Insurance
2019-01-01 open Expenses:Non-Retirement:Auto:Fit:Gas
2019-01-01 * "Transaction"
  Expenses:Non-Retirement:Auto:Fit:Insurance      10 USD
  Expenses:Non-Retirement:Auto:Fit:Gas            20 USD
  Assets:Transfer
2021-06-17 open Expenses:Non-Retirement:Auto:Fit
2021-06-17 close Expenses:Non-Retirement:Auto:Fit
```
Note that this this generates an error, because you have the open and the close directive on the same day, and this is not currently allowed. Let's ignore this for the moment. The `beancount.plugins.close_tree` adds the missing `close` directives based on the `open` directives present in the ledger:
```
2019-01-01 open Assets:Transfer
2019-01-01 open Expenses:Non-Retirement:Auto:Fit:Insurance
2019-01-01 open Expenses:Non-Retirement:Auto:Fit:Gas
2019-01-01 * "Transaction"
  Expenses:Non-Retirement:Auto:Fit:Insurance      10 USD
  Expenses:Non-Retirement:Auto:Fit:Gas            20 USD
  Assets:Transfer
2021-06-17 open Expenses:Non-Retirement:Auto:Fit
2021-06-17 close Expenses:Non-Retirement:Auto:Fit
2021-06-17 close Expenses:Non-Retirement:Auto:Fit:Insurance
2021-06-17 close Expenses:Non-Retirement:Auto:Fit:Gas
```
What you want to do is to remove the pair of `open` and `close` directives that appear on the same day. But waht if there is also a transaction to the relative account on the same day? You would need to keep the `open` and `close` directives. I think that the error message for `open` and `close` on the same day is not correct: Beancount puts `open` directives at the beginning of day and `close` directives at the end of day, and transactions in between. Thus I don't see a problem with `open` and `close` on the same day.

Reply:
1. Fixed Beancount's error message.
2. Do we need to remove the pair, if there wasn't a transaction found to the account in question on that date? We would save the potentially expensive cost of checking if we inserted the `open`/`close` pair (and sometimes ended up with this additional pair).